Refactor PGP handling and streamline encryption process.

Spomky opened the pull request:
Hi @the-pulli,

I reorganized PGP-related classes into a cleaner structure, introducing `PgpMime`, `PgpProcess`, `PgpEncrypter`, and `PgpSigner`. Removed unused traits and dependencies like `pear/crypt_gpg` in favor of a native `Process`-based implementation for better maintainability and efficiency.
My mail provider is ProtonMail and I successfully received signed, encrypted, signed and encrypted emails.
When the signing public key is provided, it is attached to the email.

#
**⚠ Warning**: Please note that:
* I did not changed the tests thus they will undoubtedly fail.
* I did not tried with passphrase protected keys. TBC

I hope this will help you with the PR.

* Encrypted
![2025-02-19_15h57_05](https://github.com/user-attachments/assets/767e08e6-c43c-433e-8329-577f4d885f0f)

* Signed
![2025-02-19_15h56_15](https://github.com/user-attachments/assets/2393ed02-d4bc-4d20-a959-6d57a02b074b)

* Signed and Encrypted
![2025-02-19_15h56_04](https://github.com/user-attachments/assets/51aa94bb-7ee7-41b6-8e66-c5c883faf731)


How to use:

```php
$signer = new \Symfony\Component\MimePgp\PgpSigner(
    __DIR__ . '/private.asc',
    __DIR__ . '/public.asc', // option
);
$encrypter = new \Symfony\Component\MimePgp\PgpEncrypter([
    'john.doe@acme.com' => __DIR__. '/recipient1.asc',
]);
$email = $signer->sign($email);
$email = $encrypter->encrypt($email);

$mailer->send($email);
```

Spomky commented:
Note: last commit adds passphrase support.

the-pulli commented:
Hi @Spomky,

I tested your changes this morning with ProtonMail and my local mail client MailMate.

Encryption works fine so far. But signing has some issues. As you stated Proton in their web app marks signed/signed encrypted mails correctly. Their mobile app does not:

![Bildschirmfoto 2025-02-20 um 09 25 43 jpeg](https://github.com/user-attachments/assets/1e39603b-32ab-4563-9101-7d0fa741dc2a)

And in MailMate the error is more dramatic:

![Screenshot 2025-02-20 at 08 27 17@2x](https://github.com/user-attachments/assets/ac266437-bf96-4ddb-a934-fa1c6355f92b)

As you removed the `PgpSigningTrait`, which was used in `PgpSignedPart` and also in the `sign` method of `PgpEncrypter`. You moved the implementation to the first one, but left the second one untouched. And my guess is, that's where the issue is.

I'll gonna debug this further and tinker around with your code to hopefully get the signing part working again.

And what I noticed, you can remove `pear/crypt_gpg` from the package root `composer.json` and add `symfony/process` instead 🙃

Spomky commented:
Hello,

I rebased this PR and changed the component dependencies (I completely missed it🫢).
I am not sure to clearly understand the issue with `PgpSigningTrait`. It contained only `prepareMessageForSigning` and after this refactoring, I noted this method was only used in the `PgpSigner` (I split `PgpEncrypter` into 2 different classes).
That's why I move the method and removed the trait.

My assumption is that a I may have messed up with the different parts and the their order is not compliant with the spec.

the-pulli commented:
Don't worry. I noticed it while installing it as local package 🤓

Yes, and in the former `PgpEncryper` class the trait method was also used in the `sign` method:

```php
    public function sign(Message $message, ?string $key = null, #[\SensitiveParameter] ?string $passphrase = null, bool $attachKey = false): Message
    {
        $messagePart = $message->getBody();
        $body = $messagePart->toString();

        if ($attachKey) {
            try {
                $messagePart = $this->attachPublicKey($key, $message);
            } catch (\Crypt_GPG_KeyNotFoundException $e) {
                throw new KeyNotFoundException($e->getMessage(), $e->getCode(), $e);
            } catch (\Crypt_GPG_Exception $e) {
                throw new GeneralException($e->getMessage(), $e->getCode(), $e);
            }

            $body = $messagePart->toString();
        }

        $body = $this->prepareMessageForSigning($messagePart, $body); // Here it's used
```

And I also tested the signature on Canary on iOS. There it also fails:

![Bildschirmfoto 2025-02-20 um 09 45 29](https://github.com/user-attachments/assets/248bdf54-5aea-4085-8f7d-5e5feb2ffa10)

I needed several attempts to get the signing right. It's a real pain.

Now to clarify. So to sign and encrypt a message. I first sign it and then encrypt it? Before that I used the method to sign and encrypt it in one go, so the signature get's embedded into the encrypted part and is not detached.

Spomky commented:
Yes the idea is to use the Singer or the Encrypter independently like S/Mime does.

My assumption is that the signature MIME structure is not correct.
I am reading the spec to understand what is wrong with it.
We can discuss on Slack (Symfony Devs group if you want.

Hereafter the signed message with and without the pulbic key
![2025-02-20_10h40_50](https://github.com/user-attachments/assets/40300e79-7a85-44eb-93ab-257846f31364)

![2025-02-20_10h41_16](https://github.com/user-attachments/assets/bb72b832-4de5-461d-a459-7a5551c29b5f)

the-pulli reviewed (approved):
As discussed. Approved.
